### PR TITLE
Enable tab complete for '-Format' parameter of Get-Date

### PR DESCRIPTION
**Pull request by iSazonov**
Add ArgumentCompletions attribute to '-Format' parameter of Get-Date to enable
IntelliSense (tab complete).

## Fix
1. Added ArgumentCompletions attribute to enable IntelliSense but allow not only named format values ("FileDate", "FileDateUniversal", "FileDateTime", "FileDateTimeUniversal") but also custom format strings ("YYYY-MM-dd").
2. Added a commit to replace tabs in test file.
3. Added new ArgumentCompletionsAttribute type which allow tab completions (IntelliSense) for parameters without throw on "invalid" values in contrast to ValidateSet.
4. Added tests for `-Format` parameter with named format values.
5. Added tests for ArgumentCompletionsAttribute

## Additional considerations
No tab completion test added for Get-Date because it seems we have no requirements to check tab completion for every parameter in all cmdlets.

### CODE REVIEW

**Comment by vors** [comment]
Should we use another attribute name just for the IntelliSense scenario?
`[ValidateSet(AllowAll)]` looks cumbersome. We can inherit from `ValidateSetAttribute` and use some name that doesn't mention Validation. It's also not very clean, but at least for the end user it will be not as confusing imo.

**Comment by iSazonov** [comment]
ValidateSet is a sealed.

File: src/System.Management.Automation/engine/Attributes.cs [hunk 1]
```diff
@@ -1564,6 +1564,14 @@ public sealed class ValidateSetAttribute : ValidateEnumeratedArgumentsAttribute
         private static ConcurrentDictionary<Type, IValidateSetValuesGenerator> s_ValidValuesGeneratorCache = new ConcurrentDictionary<Type, IValidateSetValuesGenerator>();
 
         /// <summary>
+        /// If AllowAll==true 'ValidateElement' always pass all elements.
+        /// Useful if we need enable IntelliSense (TabComplete) but effectively disable validation
+        /// when a parameter can accept not only the listed values but also arbitrary.
+        /// Ex.: Get-Date -Format
+        /// </summary>
+        public bool AllowAll { get; set; } = false;
```
Comment (lzybkr) [inline src/System.Management.Automation/engine/Attributes.cs]: This doesn't feel right to me - you shouldn't use a `Validation` attribute if you aren't validating.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetDateCommand.cs [hunk 2]
```diff
@@ -206,6 +206,7 @@ public int Millisecond
         /// Unix format string
         /// </summary>
         [Parameter(ParameterSetName = "net")]
+        [ValidateSetAttribute("FileDate", "FileDateUniversal", "FileDateTime", "FileDateTimeUniversal", AllowAll=true)]
```
Comment (lzybkr) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetDateCommand.cs]: I would suggest using an `ArgumentCompleter` attribute instead - it's a bit more cumbersome, but if this scenario makes sense (and I think it probably does), we can find some way to make it simpler.

**Review by lzybkr — CHANGES_REQUESTED** [review]
Thanks for thinking about the usability here - I agree we should do something here, just not like this.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetDateCommand.cs (hunk 2 above)
Comment (iSazonov) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetDateCommand.cs]: Do you suggest enhance `ArgumentCompleter` to support `[ArgumentCompleter("FileDate", "FileDateUniversal", "FileDateTime", "FileDateTimeUniversal")] `?

Comment (lzybkr) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetDateCommand.cs]: Just brainstorming here - but maybe a new attribute `ArgumentCompletions` that also implements `IArgumentCompleter` - this probably requires some additional code elsewhere, probably similar to the `ValidateSet` code for completion.

Comment (iSazonov) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetDateCommand.cs]: A new attribute requires a lot of code. It seems easiest to add a new constructor `public ArgumentCompleterAttribute(params string[] completeStrings)`

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetDateCommand.cs [hunk 3]
```diff
@@ -206,7 +206,7 @@ public int Millisecond
         /// Unix format string
         /// </summary>
         [Parameter(ParameterSetName = "net")]
-        [ValidateSetAttribute("FileDate", "FileDateUniversal", "FileDateTime", "FileDateTimeUniversal", AllowAll=true)]
+        [ArgumentCompleter("FileDate", "FileDateUniversal", "FileDateTime", "FileDateTimeUniversal")]
```
Comment (lzybkr) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetDateCommand.cs]: This is much better, but I think still a little weird, I still think it's worth considering adding a new attribute `ArgumentCompletions`. It seems like you could take your class `StringArrayArgumentCompleter`, rename it, make it an attribute, and either duplicate a little bit of logic around `ValidateSet` or `ArgumentCompleter` in the tab completion code.

Comment (iSazonov) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetDateCommand.cs]: Is your only thought that ArgumentCompleter is a strange name and you want to rename it?

Comment (lzybkr) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetDateCommand.cs]: I wouldn't rename `ArgumentCompleter` even if it is a strange name (which I suppose it is.)

It's a really minor thing, but to me, `ArgumentCompleter` implies some dynamic code to compute the completions, whereas `ArgumentCompletions` is static.

Comment (powercode) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetDateCommand.cs]: I have missed that for a long time - the ability to easily complete a set without limiting the parameter to the completions. 

Is it a reasonable compromise that you have to go IArgumentCompleter if you want to provide rich `CompletionResult`s, or should be enable this with `ArgumentCompletions` too?

Comment (iSazonov) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetDateCommand.cs]: Done next iteration.

File: src/System.Management.Automation/engine/CommandCompletion/ExtensibleCompletion.cs [hunk 4]
```diff
@@ -158,4 +158,56 @@ protected override void EndProcessing()
...
+        /// </summary>
+        public IEnumerable<CompletionResult> CompleteArgument(string commandName, string parameterName, string wordToComplete, CommandAst commandAst, IDictionary fakeBoundParameters)
+        {
+            var wordToCompletePattern = WildcardPattern.Get(string.IsNullOrWhiteSpace(wordToComplete) ? "*" : wordToComplete + "*", WildcardOptions.IgnoreCase);
+
+            foreach (var str in _completions)
+            {
+                if (wordToCompletePattern.IsMatch(str))
+                {
+                    yield return new CompletionResult(str, str, CompletionResultType.Text, str);
```
Comment (lzybkr) [inline src/System.Management.Automation/engine/CommandCompletion/ExtensibleCompletion.cs]: Should this be `CompletionResultType.ParameterValue`? I think that's the only place the attribute will have any effect.

File: src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs [hunk 5]
```diff
@@ -2002,6 +2002,26 @@ internal static string ConcatenateStringPathArguments(CommandElementAst stringAs
                 {
                 }
             }
+
+            var argumentCompletionsAttribute = parameter.CompiledAttributes.OfType<ArgumentCompletionsAttribute>().FirstOrDefault();
+            if (argumentCompletionsAttribute != null)
+            {
+                try
```
Comment (lzybkr) [inline src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs]: You are calling internal code which doesn't throw, so the try/catch is unnecessary.

File: src/System.Management.Automation/engine/Attributes.cs [hunk 6]
```diff
@@ -1625,6 +1633,11 @@ protected override void ValidateElement(object element)
                             Metadata.ValidateNotNullFailure);
             }
 
+            if (AllowAll == true)
```
Comment (lzybkr) [inline src/System.Management.Automation/engine/Attributes.cs]: This should be reverted.

File: src/System.Management.Automation/engine/Attributes.cs (hunk 1 above)
Comment (lzybkr) [reply src/System.Management.Automation/engine/Attributes.cs]: This should be reverted.

File: src/System.Management.Automation/engine/parser/Compiler.cs [hunk 7]
```diff
@@ -1336,6 +1336,10 @@ private static Attribute NewValidateSetAttribute(AttributeAst ast)
                 {
                     result.IgnoreCase = s_attrArgToBoolConverter.Target(s_attrArgToBoolConverter, argValue);
                 }
+                else if (argumentName.Equals("AllowAll", StringComparison.OrdinalIgnoreCase))
```
Comment (lzybkr) [inline src/System.Management.Automation/engine/parser/Compiler.cs]: This should be reverted.

**Review by lzybkr — CHANGES_REQUESTED** [review]
Just a little cleanup and I think it looks good.

File: src/System.Management.Automation/engine/CommandCompletion/ExtensibleCompletion.cs (hunk 4 above)
Comment (iSazonov) [reply src/System.Management.Automation/engine/CommandCompletion/ExtensibleCompletion.cs]: I agree. 
Fixed.

File: src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs (hunk 5 above)
Comment (iSazonov) [reply src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs]: Fixed.

File: src/System.Management.Automation/engine/Attributes.cs (hunk 6 above)
Comment (iSazonov) [reply src/System.Management.Automation/engine/Attributes.cs]: Fixed.

File: src/System.Management.Automation/engine/parser/Compiler.cs (hunk 7 above)
Comment (iSazonov) [reply src/System.Management.Automation/engine/parser/Compiler.cs]: Fixed.

**Review by lzybkr — APPROVED** [review]
Code changes are approved - but we should get a couple of tests for the new `ArgumentCompletions` attribute.

**Comment by daxian-dbw** [comment]
@iSazonov Can you please update the PR description? Looks like the design has changed.

**Comment by iSazonov** [comment]
Could you please help:  C# test failed on "ArgumentCompleter" abbreviation - full name ArgumentCompleterAttribute works. I cannot find how make "ArgumentCompleter" alias.

File: test/powershell/Language/Parser/ExtensibleCompletion.Tests.ps1 [hunk 8]
```diff
@@ -355,3 +355,97 @@ Describe "Additional type name completion tests" -Tags "CI" {
...
+        )
+
+        $testCasesCSharp = @(
+            @{ attributeName = "ArgumentCompletions"         ; cmdletName = "Get-ArgumentCompletions"  },
+            @{ attributeName = "ArgumentCompletionsAttribute"; cmdletName = "Get-ArgumentCompletions1" }
+        )
+    }
+
+    AfterAll {
+        #Remove-Module -ModuleInfo $testModule
```
Comment (daxian-dbw) [inline test/powershell/Language/Parser/ExtensibleCompletion.Tests.ps1]: Did you forget to uncomment this?

Comment (iSazonov) [reply test/powershell/Language/Parser/ExtensibleCompletion.Tests.ps1]: Yes 😄 Today test is passed. I wonder - I don't change a code.

**Review by daxian-dbw — APPROVED** [review]

**Comment by iSazonov** [comment]
@lzybkr Thanks for great idea!

**Pull request merged**